File: prototype/Assets/microcosmicWar/Scripts/zz/zzCreatorUtility.js
//统一单机与联网的创建与销毁接口


class IzzGenericCreator
{
	virtual function Instantiate(prefab : Object, position : Vector3, rotation : Quaternion, group : int) : Object
	{
	}
	
	virtual function Destroy (gameObject : GameObject):void
	{
	}
	
	virtual function isMine(networkView:NetworkView)
	{
		return true;
	}
	
	virtual function sendMessage(gameObject:GameObject,methodName : String )
	{
	}
	
	virtual function sendMessage(gameObject:GameObject,methodName : String, value : Object)
	{
	}
	
}

class ZzNetCreator extends IzzGenericCreator
{
	virtual function Instantiate(prefab : Object, position : Vector3, rotation : Quaternion, group : int) 
	{
		return Network.Instantiate(prefab, position, rotation, group);
	}
	
	virtual function Destroy (gameObject : GameObject)
	{
		Network.Destroy(gameObject);
	}
	virtual function isMine(networkView:NetworkView)
	{
		return networkView.isMine;
	}
	
	virtual function sendMessage(gameObject:GameObject,methodName : String)
	{
		gameObject.networkView.RPC(methodName,RPCMode.All); 
	}
	
	virtual function sendMessage(gameObject:GameObject,methodName : String, value : Object)
	{
		gameObject.networkView.RPC(methodName,
			RPCMode.AllBuffered,
			value); 
	}
}

class ZzSingleCreator extends IzzGenericCreator
{
	virtual function Instantiate(prefab : Object, position : Vector3, rotation : Quaternion, group : int) 
	{
		return prefab.Instantiate(prefab, position, rotation);
	}
	
	virtual function Destroy (gameObject : GameObject)
	{
		gameObject.Destroy(gameObject);
	}
	
	virtual function isMine(networkView:NetworkView)
	{
		return true;
	}
	
	virtual function sendMessage(gameObject:GameObject,methodName : String)
	{
		gameObject.SendMessage(methodName);
	}
	
	virtual function sendMessage(gameObject:GameObject,methodName : String, value : Object)
	{
		gameObject.SendMessage(methodName,value);
	}
}

//默认为单机,省去初始化操作
static var zzGenericCreator:IzzGenericCreator=ZzSingleCreator();

static var host=true;

static function resetCreator()
{
	if(Network.peerType ==NetworkPeerType.Disconnected)
	{
		//print("Network.peerType ==NetworkPeerType.Disconnected");
		zzGenericCreator=ZzSingleCreator();
		host=true;
	}
	else
	{
		//print("Network.peerType !=NetworkPeerType.Disconnected");
		zzGenericCreator=ZzNetCreator();
		if (Network.isServer)
			host=true;
		else
			host=false;
	}
	//print(zzGenericCreator);
	//zzGenericCreator
	//print(host);
}

static function Instantiate(prefab : Object, position : Vector3, rotation : Quaternion, group : int) 
{
	//Debug.Log("Instantiate");
	//Debug.Log(zzGenericCreator);
	return zzGenericCreator.Instantiate(prefab, position, rotation, group);
}
static function Destroy (gameObject : GameObject)
{
	zzGenericCreator.Destroy(gameObject);
}

//单机时与作为服务器时返回真
static function isHost()
{
	return host;
}

static function isMine(networkView:NetworkView)
{
	return zzGenericCreator.isMine(networkView);
}

//在单机中避免RPC
static function sendMessage(gameObject:GameObject,methodName : String, value : Object)
{
	zzGenericCreator.sendMessage(gameObject,methodName,value);
}

static function sendMessage(gameObject:GameObject,methodName : String)
{
	zzGenericCreator.sendMessage(gameObject,methodName);
}

//联网时使用netMethodName,单机时 使用 singleMethodName
static function sendMessag2Two(gameObject:GameObject,singleMethodName : String, netMethodName : String,value : Object)
{
	if(Network.peerType ==NetworkPeerType.Disconnected)
		gameObject.SendMessage(singleMethodName,value);
	else
		gameObject.networkView.RPC(netMethodName,
			RPCMode.AllBuffered,
			value); 
}

//function Update () {
//}

File: prototype/Assets/microcosmicWar/Scripts/zz/zzCreatorUtility.cs
using UnityEngine;
using System.Collections;

public class zzCreatorUtility : MonoBehaviour
{

    //统一单机与联网的创建与销毁接口


    abstract class IzzGenericCreator
    {
        public abstract GameObject Instantiate(GameObject prefab, Vector3 position, Quaternion rotation, int group);
        //{
        //    Debug.LogError("IzzGenericCreator.Instantiate");
        //}

        public abstract void Destroy(GameObject gameObject);
        //{
        //    Debug.LogError("IzzGenericCreator.Destroy");
        //}

        public abstract bool isMine(NetworkView networkView);
        //{
        //    Debug.LogError("IzzGenericCreator.isMine");
        //    return true;
        //}

        public abstract void sendMessage(GameObject gameObject, string methodName);
        //{
        //    Debug.LogError("IzzGenericCreator.sendMessage");
        //}

        public abstract void sendMessage(GameObject gameObject, string methodName, object value);
        //{
        //    Debug.LogError("IzzGenericCreator.sendMessage");
        //}

    }

    class ZzNetCreator : IzzGenericCreator
    {
        public override GameObject Instantiate(GameObject prefab, Vector3 position, Quaternion rotation, int group)
        {
            //Debug.Log("Network.Instantiate"+prefab.name);
            return Network.Instantiate(prefab, position, rotation, group) as GameObject;
        }

        public override void Destroy(GameObject gameObject)
        {
            Network.Destroy(gameObject);
        }
        public override bool isMine(NetworkView networkView)
        {
            return networkView.isMine;
        }

        public override void sendMessage(GameObject gameObject, string methodName)
        {
            gameObject.networkView.RPC(methodName, RPCMode.All);
        }

        public override void sendMessage(GameObject gameObject, string methodName, object value)
        {
            gameObject.networkView.RPC(methodName,
                RPCMode.All,
                value);
        }
    }

    class ZzSingleCreator : IzzGenericCreator
    {
        public override GameObject Instantiate(GameObject prefab, Vector3 position, Quaternion rotation, int group)
        {
            //Debug.Log("ZzSingleCreator.Instantiate");
            return GameObject.Instantiate(prefab, position, rotation) as GameObject;
        }

        public override void Destroy(GameObject gameObject)
        {
            GameObject.Destroy(gameObject);
        }

        public override bool isMine(NetworkView networkView)
        {
            return true;
        }

        public override void sendMessage(GameObject gameObject, string methodName)
        {
            gameObject.SendMessage(methodName);
        }

        public override void sendMessage(GameObject gameObject, string methodName, object value)
        {
            gameObject.SendMessage(methodName, value);
        }
    }

    //默认为单机,省去初始化操作
    static IzzGenericCreator zzGenericCreator = new ZzSingleCreator();

    static bool host = true;

    public static void resetCreator(zzCreatorUtility pCreator)
    {
        if (Network.peerType == NetworkPeerType.Disconnected)
        {
            //print("Network.peerType ==NetworkPeerType.Disconnected");
            zzGenericCreator = new ZzSingleCreator();
            host = true;
        }
        else
        {
            //print("Network.peerType !=NetworkPeerType.Disconnected");
            pCreator.inNetwork = true;
            zzGenericCreator = new ZzNetCreator();
            if (Network.isServer)
            {
                pCreator.isServer = true;
                host = true;
            }
            else
                host = false;
        }
        //Debug.Log(zzGenericCreator.isMine(null));
        //zzGenericCreator
        //print(host);
    }

    public static GameObject Instantiate(GameObject prefab, Vector3 position, Quaternion rotation, int group)
    {
        //Debug.Log("Instantiate");
        //Debug.Log(zzGenericCreator);
        return zzGenericCreator.Instantiate(prefab, position, rotation, group);
    }
    public static void Destroy(GameObject gameObject)
    {
        zzGenericCreator.Destroy(gameObject);
    }

    //单机时与作为服务器时返回真
    public static bool isHost()
    {
        return host;
    }

    public static bool isMine(NetworkView networkView)
    {
        return zzGenericCreator.isMine(networkView);
    }

    //在单机中避免RPC
    public static void sendMessage(GameObject gameObject, string methodName, object value)
    {
        zzGenericCreator.sendMessage(gameObject, methodName, value);
    }

    public static void sendMessage(GameObject gameObject, string methodName)
    {
        zzGenericCreator.sendMessage(gameObject, methodName);
    }

    //联网时使用netMethodName,单机时 使用 singleMethodName
    public static void sendMessag2Two(GameObject gameObject, string singleMethodName, string netMethodName, Object value)
    {
        if (Network.peerType == NetworkPeerType.Disconnected)
            gameObject.SendMessage(singleMethodName, value);
        else
            gameObject.networkView.RPC(netMethodName,
                RPCMode.All,
                value);
    }

    void Awake()
    {
        //Debug.LogError("zzCreatorUtility.Awake");
        resetCreator(this);
    }

    public bool inNetwork = false;
    public bool isServer = false;
    //void  Update (){
    //}
}